File: solution/846. Hand of Straights/846.cpp
class Solution {
public:
    bool isNStraightHand(vector<int>& hand, int groupSize) {
        if (groupSize == 1) return true;
        sort(hand.begin(), hand.end());
        
        unordered_map<int, queue<int>> m;
        int left = hand.size();
        for (int card : hand) {
            int pre = card - 1;
            if (m.count(pre)) {
                int cnt = m[pre].front() + 1; m[pre].pop();
                if (cnt < groupSize)
                    m[card].push(cnt);
                else
                    left -= cnt;
                
                if (m[pre].empty())
                    m.erase(pre);
            } else {
                m[card].push(1);
            }
        }
        return left == 0;
    }
};

File: solution/846. Hand of Straights/846.py
class Solution:
    def isNStraightHand(self, hand: List[int], groupSize: int) -> bool:
        m = Counter(hand)
        for card in sorted(m):
            if m[card] > 0:
                for nxt in range(card + 1, card + groupSize):
                    m[nxt] -= m[card]
                    if m[nxt] < 0:
                        return False
                m[card] = 0
        return True
    
